### ISSUE TITLE
Advice speeding up autograd for loss functions with expensive inner loop?

### ISSUE THREAD

**Issue by michaelchughes** [HIPS/autograd#157]
I have autograd working on a loss function for a pretty large scale problem... essentially a version of topic models for text documents. All seems to be working fine, except the performance seems to be a bit too slow for my application: on toy data, if I compute the loss function in X seconds for 100 documents, evaluating the gradient can take about 25-100X seconds. Rumor has it that usually, 2-3X might be the expected cost, so I'm wondering:

* (1) is 25X slowness unusual? or is it ok?
* (2) any advice on writing code differently to speed things up?

Below is a minimal "non-working" example to give the flavor of the problem. The overall loss function works out to a sum over documents, where at each document we perform many iterations (typically 25-100) of an inner loop that iteratively estimates a needed vector (more iterates = higher quality) . Every line of code is using vectorized operations as much as possible. The truly expensive part based on my profiling is just that the inner while loop at each document requires many iterations.

```
def calc_loss_for_many_docs(topics_KV):
    loss = 0.0
    for d in xrange(n_docs):
        vector_d_K = calc_vector_for_single_doc(topics_KV[:, specific_words_d_U])
        loss = loss - inner(weights_K, vector_d_K)
    return loss

def calc_vector_for_single_doc(topics_KU):
    ct_topics_KU = topics_KU * word_ct_d_U[np.newaxis, :]
    K = topics_KU.shape[0]
    pi_d_K = np.ones(K) / float(K)
    # Start loop over iterations
    giter = 0
    while giter < max_iters:
        giter = giter + 1
        denom_U = 1.0 / np.dot(pi_d_K, topics_KU)
        grad_K = pi_step_size * np.dot(ct_topics_KU, denom_U)
        pi_d_K = pi_d_K * np.exp(grad_K)
        pi_d_K_sum = np.sum(pi_d_K)
        pi_d_K = pi_d_K / pi_d_K_sum
    return pi_d_K
```

I'm happy to give more details and real working code examples if that would be helpful, but I thought at first I'd just generally ask for quick ideas about this. My other experience with Python leads me to believe that the inner while loop is always a bad sign and that moving to Cython/C++ is needed. But I don't want to lose autograd magic that way.

Thanks!

**Comment by michaelchughes** [comment]
For reference, here's some careful timings, looking at how computation time scales vs. the number of iterations (max_iters) done in the inner loop. It's sensible that as max_iters doubles, so does the median run time (of 3 reps/trials). 

I seem to get pretty consistent ~25X scaling between the runtimes of the loss and the gradient. Perhaps this is just to be expected?

```
--->>> opt_kwargs: max_iters=10
n_reps 3  median sec    0.0673  max sec    0.0700  loss_with_cur_max_iter
n_reps 3  median sec    1.6843  max sec    1.7370  gradient_loss_with_cur_max_iter_wrt_argnum_0
--->>> opt_kwargs: max_iters=20
n_reps 3  median sec    0.1276  max sec    0.1313  loss_with_cur_max_iter
n_reps 3  median sec    3.3237  max sec    3.3664  gradient_loss_with_cur_max_iter_wrt_argnum_0
--->>> opt_kwargs: max_iters=40
n_reps 3  median sec    0.2429  max sec    0.2510  loss_with_cur_max_iter
n_reps 3  median sec    6.5577  max sec    7.0215  gradient_loss_with_cur_max_iter_wrt_argnum_0
--->>> opt_kwargs: max_iters=80
n_reps 3  median sec    0.4889  max sec    0.5119  loss_with_cur_max_iter
n_reps 3  median sec   11.9420  max sec   21.3206  gradient_loss_with_cur_max_iter_wrt_argnum_0
```

**Comment by kcarnold** [comment]
A few ideas:
* Do you need the iterative fit to happen for every evaluation of the cost function? That seems like it's going to be expensive and potentially inaccurate regardless of implementation. You might be able to reparameterize the problem with a trick like the Augmented Lagrangian method -- someone more skilled in optimization could speak to that.
* You don't have to lose autograd entirely if you change some of the implementation -- see https://github.com/HIPS/autograd/blob/master/examples/define_gradient.py. Even hand-coding the gradient of one iteration in pure NumPy could help if you can make some simplifications or optimize memory use.
* Since the loss is a sum over documents, so is the gradient. You can save memory for intermediate values by running autograd on only one document and summing the gradients. I doubt this will help if you're purely compute-bound, though.

**Comment by mattjj** [comment]
Great to hear from you guys!

That's definitely a much bigger slowdown than we would like. Ken's suggestions are good ones. A general principle to keep in mind is that autograd works at the level of numpy/scipy functions as primitives, and that means the same performance rule of thumb to "do as much as possible in terms of numpy calls" is even more important for autograd code, since Python overheads can get amplified with autograd on both forward and backward passes while numpy calls stay cheap. That happens because autograd is tracing how primitive (numpy) operations are getting composed, then acting kind of like an interpreter (implemented in Python) on those traced operations during the reverse pass.

It could be that numpy and scipy don't provide the right primitives. In that case it can be helpful to define your own primitives and, as Ken suggests, define their gradients (ideally in terms of other primitives, so that you can get higher-order differentiation if you need it). If you can define part or all of your inner loop body as a new primitive, you don't have to lose autograd magic, and it could be much faster because it would eliminate a lot of the overhead that autograd is probably amplifying by tracing all the steps of all the iterations of that loop body, even if your primitive and its gradient are written entirely in Python.

To get the absolute best performance, you could push individual primitives down into Cython; I've been writing some [snippets to do that with variants of scipy.linalg funcions](https://github.com/mattjj/autograd_linalg). However, since your forward pass is relatively fast, I think the thing to tackle first is reducing autograd overhead. Defining your loop body as a primitive and defining its gradient could be the most direct path to doing that, and you can always further optimize by pushing things into Cython afterwards if you need to.

Your code might end up looking something like this:

```python
from autograd.core import primitive

# [...]

def calc_vector_for_single_doc(topics_KU):
    ct_topics_KU = topics_KU * word_ct_d_U[np.newaxis, :]
    K = topics_KU.shape[0]
    pi_d_K = np.ones(K) / float(K)
    # Start loop over iterations
    giter = 0
    while giter < max_iters:
        giter = giter + 1
        pi_d_K = update_pi(pi_d_K, ct_topics_KU, topics_KU)
    return pi_d_K

@primitive
def update_pi(pi_d_K, ct_topics_KU, topics_KU):
    denom_U = 1.0 / np.dot(pi_d_K, topics_KU)
    grad_K = pi_step_size * np.dot(ct_topics_KU, denom_U)
    pi_d_K = pi_d_K * np.exp(grad_K)
    pi_d_K_sum = np.sum(pi_d_K)
    return pi_d_K / pi_d_K_sum

update_pi.defgrad(lambda updated_pi_d_K, pi_d_K, ct_topics_KU, topics_KU:
                  lambda g: ...,  # TODO my cool Jacobian-transpose operation
                  argnum=0)
```

Where `g` is what we tend to call the 'outgrad', which is ∂ loss / ∂ updated_pi_d_K. One important thing to figure out is whether you'd need to define gradients with respect to the other arguments, `ct_topics_KU` and `topics_KU`. You can determine that by testing if those variables in your current code come in as boxed autograd `Node` types.

Here are some orthogonal thoughts. It looks like your inner loop is a fixed-stepsize steepest ascent optimization. Do you think it typically converges, up to some reasonable numerical tolerance? If so, there are two tweaks that could help without much code. First, you could test for convergence within some tolerance and break out of the loop early (since the average number of steps required to converge may be much less than the maximum). That could avoid superfluous loop iterations, where the overhead gets amplified by autograd. Second, if you run an inner optimization to a fixed point, you may not need to propagate gradients through the optimization procedure itself; you may just need to use the optimizing value of the decision variable and apply the implicit function theorem. (We used that in Appendix A of [the SVAE paper](http://arxiv.org/abs/1603.06277), though it's probably explained better elsewhere.) If that's the case, I can say more about how to do that with autograd, but basically you use the [`getval` function](https://github.com/HIPS/autograd/issues/106) (which behaves like Theano's `consider_constant` or TensorFlow's `stop_gradient`) to unbox the values while you're computing the fixed point.

**Comment by mattjj** [comment]
@kcarnold re: the augmented Lagrangian idea, did you mean you might treat the first-order optimality conditions of the inner problem as constraints, then solve the resulting overall problem using an augmented Lagrangian method?

**Comment by kcarnold** [comment]
Exactly. But your point about the implicit function theorem is probably a better idea.

**Comment by michaelchughes** [comment]
@mattjj Thanks for the detailed response! I'll take a look at defining my own primitive for the inner loop content, and will report back soon-ish if things work out.

RE your orthogonal thoughts:
1) I do have an "early convergence detection" already working in my real application. It's a good suggestion and definitely reduces overhead. But still, usually the loop needs 40-100 iterations before convergence occurs, so there's still plenty of cost yet to save.
2) Avoiding the gradient entirely would be nice, but doesn't seem to apply in our case. Had a careful discussion with Andy Miller about it. Still not 100% sure (if there's a more simple/clear presentation of those ideas, that would be awesome), but looks like we need to take gradients all the way through. Bummer.
